File: reference/sealed.py
"""Sealed-ballot packing, key derivation and encryption (spec B4)."""

import grumpkin
import poseidon2
from keccak import keccak256

from ballots import pack, unpack, validate  # noqa: F401  (re-exported for the Noir tests)

DOMAIN = int.from_bytes(b"RankedShares/sealed/v2", "big")


def derive_sk(master, key_salt):
    return int.from_bytes(keccak256(master + key_salt), "big") % grumpkin.Q


def pubkey(sk):
    return grumpkin.mul(sk, grumpkin.G)


def mask(shared, voter):
    return poseidon2.hash([DOMAIN, shared[0], shared[1], voter])


def encrypt(pk, voter, ranks, k):
    r = grumpkin.mul(k, grumpkin.G)
    s = grumpkin.mul(k, pk)
    c = (pack(ranks) + mask(s, voter)) % grumpkin.P
    return (r[0], r[1], c)


def decrypt(sk, voter, ct, m):
    rx, ry, c = ct
    if rx == 0 or not grumpkin.is_on_curve((rx, ry)):
        return None
    s = grumpkin.mul(sk, (rx, ry))
    if s is None:
        return None
    packed = (c - mask(s, voter)) % grumpkin.P
    return unpack(packed, m)

File: cre/src/lib/sealed.ts
import { DOMAIN, FIELD, Q, mod } from "./field";
import { G, isOnCurve, mul, type Point } from "./grumpkin";
import { bytesToBig, concatBytes, keccak } from "./hash";
import { sponge } from "./poseidon2";

export const MASTER_MESSAGE = "RankedShares tallier master secret v1";

export { pack, validate, unpack, effectiveRanks } from "../../../shared/ranks";
import { pack, unpack } from "../../../shared/ranks";

export function masterFromSignature(sig: Uint8Array): Uint8Array {
  return keccak(sig);
}

export function deriveSk(master: Uint8Array, keySalt: Uint8Array): bigint {
  return bytesToBig(keccak(concatBytes(master, keySalt))) % Q;
}

export function mask(shared: Point, voter: bigint): bigint {
  return sponge([DOMAIN, shared.x, shared.y, voter]);
}

export function encrypt(pk: Point, voter: bigint, ranks: number[], k: bigint): [bigint, bigint, bigint] {
  const r = mul(k, G);
  const s = mul(k, pk);
  return [r.x, r.y, mod(pack(ranks) + mask(s, voter))];
}

export function decrypt(sk: bigint, voter: bigint, ct: [bigint, bigint, bigint], m: number): number[] | null {
  const [rx, ry, c] = ct;
  if (rx === 0n || !isOnCurve(rx, ry)) return null;
  const s = mul(sk, { x: rx, y: ry });
  return unpack(mod(c - mask(s, voter)), m);
}
